File: src/example/scala/ContentionWatcher.scala
import org.jdiscript.JDIScript
import org.jdiscript.util.VMSocketAttacher

import org.jdiscript.HandlerConversions._
import com.sun.jdi.event._

object ContentionWatcher {
  def main(args: Array[String]): Unit = {
    val vm = new VMSocketAttacher(12345).attach
    val j = new JDIScript(vm)

    j.monitorContendedEnterRequest((ev: MonitorContendedEnterEvent) => {
      j.printTrace(ev, "ContendedEnter for " + ev.monitor)
    }).enable

    j.monitorContendedEnteredRequest((ev: MonitorContendedEnteredEvent) => {
      val ts = System.currentTimeMillis
      println("%s: %s: ContendedEntered for %s".format(ts, ev.thread, ev.monitor))
    }).enable

    j.run((ev: VMStartEvent) => { println("Got StartEvent") })

    println("Shutting down")
  }
}

File: src/example/clojure/ContentionWatcher.clj
(import org.jdiscript.JDIScript
        org.jdiscript.util.VMSocketAttacher
        (org.jdiscript.handlers OnMonitorContendedEnter
                                OnMonitorContendedEntered
                                OnVMStart)
        com.sun.jdi.event.MonitorContendedEnteredEvent)

(def vm (.attach (new VMSocketAttacher 12345)))
(def j (new JDIScript vm))

(.enable (.monitorContendedEnterRequest j
  (proxy [OnMonitorContendedEnter] []
    (monitorContendedEnter [ev]
      (.printTrace j ev (str "ContendedEnter for " (.monitor ev)))))))

(.enable (.monitorContendedEnteredRequest j
  (proxy [OnMonitorContendedEntered] []
    ; Type hint needed to avoid "Can't call public method of non-public class"
    (monitorContendedEntered [^MonitorContendedEnteredEvent ev]
      (let [timestamp (System/currentTimeMillis)]
        (println (str timestamp ": " (.thread ev)
                    ": ContendedEnter for " (.monitor ev))))))))

(.run j
  (proxy [OnVMStart] []
    (vmStart [ev] (println "Got StartEvent"))))


(println "Shutting Down")